Content for the solution pattern (#1)
cardil/solution-pattern-event-mesh-for-microservices: documentation/modules/ROOT/pages/02-architecture.adoc

= Building Apps around the Event Mesh
:sectnums:
:sectlinks:
:doctype: book

= Architecture 

The architecture of an event mesh-enabled system is a paradigm shift from traditional transactional designs to an eventual consistency model.
This design aligns better with real-world processes, where different parts of a system may operate asynchronously yet collaboratively.

In this section, we explore the technologies, flow, and structure that make the event mesh architecture resilient, scalable, and developer-friendly.

[#tech_stack]
== Technology Stack

Here's the list of technologies used in this solution and its examples:

* Red Hat supported products
** https://www.redhat.com/en/technologies/cloud-computing/openshift[Red Hat OpenShift]
&mdash; Orchestrate containerized applications.

> Based on https://kubernetes.io/[Kuberentes].

Reviewer: ```suggestion
Based on https://kubernetes.io/[Kubernetes].
```

** https://www.redhat.com/en/technologies/cloud-computing/openshift/serverless[Red Hat OpenShift Serverless]
&mdash; Provides the _Event Mesh_ and _Serverless_ capabilities.
Based on https://knative.dev[Knative] project.

> ** https://swc.saas.ibm.com/en-us/redhat-marketplace/products/red-hat-amq[Red Hat AMQ Streams]

Reviewer 2: I think the name of the product is now "Streams for Apache Kafka": https://developers.redhat.com/products/streams-for-apache-kafka/overview

Author: Yes, I was thinking the same, but this is so confusing to me. 

I read somewhere it's, in fact, the name of the managed Kafka solution, but it looks it's aligning to replace the name. That's why I put IBM page there.

Anyway, I'll replace the reference with _Streams for Apache Kafka_ and link you provided.

&mdash; (Optional) Provides a persistence for _Event Mesh_, likely needed in production.
Based on https://strimzi.io/[Strimzi] project.
* Other open source technologies:

> ** https://cloudevents.io/[CloudEvents] &mdash; Provides a standard for event metadata

Reviewer 2: Perhaps in this list also CNCF CloudEvents

Author: IDK, if it's a good idea to add CNCF to all those links...

- CNCF Strimzi
- CNCF Knative
- CNCF Cloud Events
- CNCF OpenTelemetry

Isn't that too mouthful for you?

Reviewer 2: That's true :-) 

I'd perhaps just say that once (for the first mention) for CloudEvents. Why? I believe part here is also promoting it and underlining its standard characteristic is IMO a good idea.

Author: I added CNCF note on line 33

** https://opentelemetry.io/[OpenTelemetry] &mdash; (Optional) Facilitates tracing for observability.

[#in_depth]
== An in-depth look at the solution's architecture

Building applications around the Event Mesh is a solution that can be applied both to existing and new projects.
For existing applications, domain boundaries can guide the division into modular components, which may evolve into separate services.
These modules will generate commands intended for the _Event Mesh_, which will then route these events to their designated endpoints.
The _Event Mesh_ will allow for such transition, gradually making the system not only more responsive but also better suited to the real-world business logic.

=== Problem Analysis

Traditional systems often enforce strict transactional consistency, which can impede application performance and compromise business logic.
For instance, upon the completion of a ride-sharing service at the end-user's destination, the system should reliably capture this real-world event.
The capture should be performed regardless of any potential operational disruptions affecting dependent services (e.g., invoicing).

> In such scenarios, transactional applications typically return an error, which prevents any data from being changed, and causes the loss of real-world intent of end-users.

Reviewer: You could add here also, that distributed transactions typically embrace a number of steps in a user workflow which are all undone, causing frustration for users who have to redo everything, not ally the part that didn't work. (just a suggestion)

Author: Yes, this was my intention. And I think I wrote about it somewhere, but later maybe removed it. I agree, this could be a good idea to explain the problem a bit more clearly!

Author: Found it. I know I wrote about it somewhere.

It is on the demo page, here: https://github.com/cardil/solution-pattern-event-mesh-for-microservices/blob/66b1614e2da878c966f023bdd06ed9433742c638/documentation/modules/ROOT/pages/03-demo.adoc?plain=1#L73

Maybe in fact it would be better to move it out of that page and put it here...

This results in an adverse user experience and a deviation from the genuine business process, potentially leading to customer dissatisfaction.

=== Solution Breakdown

The concept of employing the _Event Mesh_ as a central, reliable hub for dispatching commands, as events, lies at the heart of this solution.
This approach aligns closely with the Command Query Responsibility Segregation (_CQRS_) pattern, which distinctly categorizes commands and queries.
Commands, in this context, are modeled as asynchronous events, designed to undergo processing by the _Event Mesh_.
On the other hand, queries are synchronous operations, safe to retry, ensuring no loss of data integrity due to transient errors.

The primary responsibility of the _Event Mesh_ is twofold.
Firstly, it persists the incoming events, thereby maintaining a record of changes in the system's state.
Secondly, it routes these events to their respective endpoints, ensuring that the appropriate microservices are notified and can subsequently update their internal states based on the event data.

> The mesh's inherent resilience is further bolstered by its exponential backoff strategy, which it employs when encountering operational failures.

Reviewer: why is an _exponential_ backoff strategy benefitial and not only one that does retries after a fixed amount of time ? This could be added here, as otherwise I would leave out _exponential backoff_ and replace it with just retries to keep it more simple.

Author: Maybe I could add a sentence or two about exponential backoff - it gives the overwhelmed endpoint time to recover, which in many cases allow system to recover without human intervention.

Reviewer 2: I think just retries is good enough at this level.

linear / exponential are more implementation details - which have their pros/cons. We could (later?) link to the knative page: https://knative.dev/docs/eventing/event-delivery/#configuring-subscription-event-delivery  but it also does not really well explain _why_ for each

This mechanism ensures that the system retries the operation until it succeeds, thus mitigating the risk of data loss or system disruption due to transient issues.

> By integrating the _Event Mesh_ into the system architecture, several architectural benefits are achieved:

Reviewer 2: I think in the list below I am missing a sentence on the fact that there is no _direct_ communication or invocation between the implemented applications / services. However they are loosely coupled and communicate only via events.

Author: I'm sure it's mentioned somewhere else on the page (maybe another?!?). But yes, we surely can add that point.

Author: It's written in just the next point.

> * **Decomposition of the application into independently functioning services**:

Reviewer: In this section you could drop "Domain Driven Design" somehow as well.

Author: Do you think I could just drop-in "Domain Driven Design" here, without explaining what it is?

Reviewer: You could drop it again with a red hat internal link. Pretty sure we have something about DDD and Microservices. If not, also ok :)

Author: Maybe this page: https://en.wikipedia.org/wiki/Domain-driven_design

Author: Added, see if fits

This approach facilitates a division of labor, with each service handling specific responsibilities.
This not only enhances maintainability but also fosters scalability, as services can be independently scaled based on their demands.

* **Improved business alignment**:

> By embracing an eventual consistency model, the _Event Mesh_ aligns closely with the inherent nature of most real-world business processes.

Reviewer 2: This sounds very "teachy" :man_shrugging:

Author: The point of this was to show the eventual consistency fits better to the majority of business cases, instead of transactional approach. The transactional approach is widely miss-used, de-facto changing the business processes to force transactional behavior when it doesn't fit, for ex.: rollback of transaction when the invoicing application is unresponsive.

Do you think we can re-write to sound less "teachy" and still push the narrative?

Unlike traditional transactional systems that strive for immediate, irreversible consistency, the _Event Mesh_ allows for a more flexible and adaptive approach to data consistency.
This results in better alignment with business requirements, as it supports scenarios where multiple services collaborate and synchronize their operations, making the whole state eventually consistent, without the constraint of strict, synchronous consistency.

> * **Avoidance of systemic issues**:

Reviewer: ```suggestion
* **Improved Resilience**:
```

> The _Event Mesh_'s error-handling mechanism, through retries and event persistence, aids in minimizing the impact of failures on the end user.

Reviewer 2: We also have "dead letter sink" that can be optionally configured. 

so, even if a "consumer" is unable to receive the event. It will be routed to the DLS.

Now, if the DLS is also a Knative Broker, the benefit is that all undelivered events can be consumed from there, using the standard Knative Broker and Trigger APIs.

Author: I was reluctant to introduce the DLS to not complicate the solution even further. Maybe I can put some hint about it, without introducing it fully?!?

Reviewer 2: > I was reluctant to introduce the DLS to not complicate the solution even further.

I see your point. Maybe we can just hint to the concept, and link to the docs as "out of scope" on _this_ doc? I geuss that would be fine, and still "complete"

This is crucial as it prevents bugs and outages from becoming visible to the user, thereby preserving the system's perceived responsiveness.

* **Enhanced system performance**:
The system becomes more responsive, as the end user no longer needs to wait for multiple, often independent, operations to complete successfully.
The _Event Mesh_'s event-driven model, coupled with the retries and event persistence, ensures that critical state changes are propagated swiftly and reliably, thereby improving the overall user experience.

=== _Event Mesh_ Flow

The event-driven flow enables eventual consistent collaboration and state synchronization between services, fostering a resilient, scalable, and developer-friendly system architecture.

A usual flow may look like:

> 1. An end-user application sends a request, which forms a _Command_ type event, which is transferred to the _Event Mesh_.

Reviewer 2: you mean app is emitting event (via HTTP) to the "mesh" ?

Author: Yep.

It's convoluted, you're saying? Maybe like this:

```suggestion
1. An end-user application sends an _HTTP_ request to the _Event Mesh_. Such message can be understood as a _Command_ type event.
```

Reviewer 2: I like this

> 2. The _Event Mesh_ (Broker) persists the event in a queue (e.g. Kafka).

Reviewer: ```suggestion
2. The _Event Mesh_ (Broker) persists the event in a queue (like a Kafka topic, but the implementation is hidden from the user).
```

Reviewer 2: I'd call it Apache Kafka (legal aspect from the ASF)

> After successful persistence _Event Mesh_ returns the success information.

Reviewer 2: there is twice "success" in this sentence. 

After ingress receives event, it does HTTP: 202 - Accepted

Author: I would like to avoid introducing additional technical terms like `ingress`. Those are not common knowledge in app-dev community, so would require additional explanation.

How about:

```suggestion
After _Event Mesh_ persists safely the data, it returns a successful _HTTP_ response with the `202 Accepted` return code.
```

Reviewer 2: perfect

At this point, the operation could already be considered successful, from the end-user point of view.
It will eventually settle correctly in all downstream systems.

> 3. The _Event Mesh_ routes the event to the appropriate endpoint based on the event's metadata and configured triggering rules.

Reviewer 2: CloudEvent metadata?

4. The endpoints receive the events and process them, updating their internal states and potentially emitting new events for downstream consumers.
The potential events are transmitted to the _Event Mesh_.
5. The dispatch loop continues until the event queue is empty and all the events are processed successfully.

> The failures are being automatically retried by the _Event Mesh_ using exponential backoff algorithm by default.

Reviewer: ```suggestion
The failures are automatically retried by the _Event Mesh_, with an increasing pauses between retries to avoid overloading the system.
```

InvoiceService ..> InvoiceDB: Update Invoice records
@enduml
////

The diagram illustrates the example flow of events between the applications, the Knative _Event Mesh_, and the datastores which persist settled state of the system.

[TIP]
====
Notice the applications aren't pulling the events from the queue.
In fact they aren't aware of any.
The _Event Mesh_ is the one controlling the flow, and retrying when needed.

There are *no additional* libraries needed to consume events from _Event Mesh_.
The _Event Mesh_ pushes the events as _CloudEvents_ encoded as _REST_ messages.
====

[NOTE]
====
The exponential backoff algorithm used by _Event Mesh_ is configurable.
It uses the following formula to calculate the backoff period: `+backoffDelay * 2^<numberOfRetries>+`, where the backoff delay is a base number of seconds, and number of retries is automatically tracked by the _Event Mesh_.

A dead letter sink can also be configured to send events in case they exceed the maximum retry number, which is also configurable. 
====

=== _Work Ledger_ analogy

A good way of thinking about the _Event Mesh_ and its persistent queue backend is the _Work Ledger_ analogy.

> Like in olden days, the clerk was keeping his to-do work in the _Work Ledger_ (e.g. a tray for paper forms).

Reviewer: ```suggestion
Like in the olden days, the clerk kept his to-do work in the _Work Ledger_ (e.g. a tray for paper forms).
```

Then he was picking the next form, and processing it, making changes within the physical file cabinets.

> In case of rare and unexpected issues (e.g. Invoicing dept being unavailable), the clerk would just put the data back onto the _Work Ledger_ to be processed later.

Reviewer: ```suggestion
In case of rare and unexpected issues (e.g. invoicing dept being unavailable), the clerk would just put the data back onto the _Work Ledger_ to be processed later.
```

When all the events are processed, the backward compatible code may be removed simplifying the maintenance.

Because in the long-term, the regular datastores are the source of truth for the system, all traditional techniques for application maintenance apply well.
It is also, easier to understand for developers, as it avoids sophisticated event handlers logic, and reconciling into the read database abstraction.

=== Differences from the _Service Mesh_

Worth pointing out are the differences from the _Service Mesh_ pattern.
The _Service Mesh_ pattern is intended to improve the resilience of synchronous communications which return the response.
The _Service Mesh_ effectively raises the uptime of the dependent endpoints by retrying and backoff policies.
The uptime can't be raised to 100%, though, so _Service Mesh_ still can lose the messages.

The table below captures the key differences:

[cols=".^1,.^2,.^2"]
|===
|| Service Mesh | Event Mesh

|
*Similarities*

2+a|

* Flexibility
* Robustness
* Decoupling

|
*Differences*
a|
* Synchronous
* Request and response
* Better for queries

a|
* Asynchronous
* Events
* Better for commands

|===

> === Supporting Legacy Systems

Reviewer 2: CloudEvent is key here - in binary mode, the exta metadata is just Headers (true for kafka/http) and actual payload stay intact .

Author: I don't follow your comment here. Can you re-state what are you proposing?

Reviewer 2: That binary cloud events are great and interoperable.

Perhaps not needed to say here

One of the strengths of an _Event Mesh_ architecture is its ability to integrate seamlessly with legacy systems, making them more resilient and adaptable.
Legacy applications can be retrofitted to produce and consume events through lightweight adapters.
For instance:

* A monolithic legacy application can send events for specific operations, instead of handling all logic internally in transactional fashion.

* Event listeners can be introduced incrementally, enabling the legacy app to subscribe to events without refactoring its core logic.

* This approach decouples old systems from rigid workflows, allowing for gradual modernization while ensuring operational continuity.

=== Improving Resilience in Applications

Traditional systems often rely on synchronous calls and transactions, which can cascade failures across components.

> Replacing these with asynchronous event-driven communication reduces dependencies and makes the system _Eventually Consistent_.

Reviewer: maybe add a link to the definition of _eventual consistency_ (like a Wikipedia page)

Author: I didn't find any RH page for eventual consistency, and IDK if we should link to other websites. Maybe that's ok?!?

Reviewer: if you don't find a good link that you can use, then maybe don't italicize but just use "eventually consistent", so that it is not really a concept that the user is not aware of.

Author: I'm going to link this: https://en.wikipedia.org/wiki/Eventual_consistency
